components/script/dom/node.rs

```diff
@@ -1401,13 +1408,22 @@ impl Node {
         let old_doc = node.owner_doc();
         // Step 2.
         node.remove_self();
+        // Step 3.
         if &*old_doc != document {
-            // Step 3.
+            // Step 3.1.
             for descendant in node.traverse_preorder() {
                 descendant.set_owner_doc(document);
             }
-            // Step 4.
             for descendant in node.traverse_preorder() {
+                // Step 3.2.
+                if let Some(element) = node.downcast::<Element>() {
+                    if element.get_custom_element_definition().is_some() {
```

Comment:
We may want to write a `Node::as_custom_element` helper which combines the downcast and custom check into one.